### PR TITLE
[Buttons] Added Shapes Support

### PR DESCRIPTION
**Pull request by yarneo**
Pivotal Story: https://www.pivotaltracker.com/story/show/156168281

* Implementation of Shapes support in Buttons.
* Added example of using shapes on different styles of buttons (see screenshot and gif below).
* Exposed an initializer for MDCCurvedCornerTreatment to be able to use in the example.
* Exposed the color sublayer in MDCShapedShadowLayer to allow hierarchal reordering of sublayers for a shaped button to work with ink and shadow correctly.
* Updated unit tests.

simulator screen shot - iphone 8 - 2018-04-16 at 18 43 40

### CODE REVIEW

**Comment by jverkoey** [comment]
Does "Stroked Triangular" intentionally not have ink?

File: components/Buttons/src/MDCButton.m [hunk 1]
```diff
@@ -860,7 +861,7 @@ - (void)updateAlphaAndBackgroundColorAnimated:(BOOL)animated {
 }
 
 - (void)updateBackgroundColor {
-  super.backgroundColor = self.currentBackgroundColor;
+  self.layer.shapedBackgroundColor = self.currentBackgroundColor;
```
Comment (jverkoey) [inline components/Buttons/src/MDCButton.m]: Should we still be setting super.backgroundColor here? I'm surprised we were doing this in a method other than setBackgroundColor.

File: components/private/ShapeLibrary/src/MDCCurvedCornerTreatment.h [hunk 2]
```diff
@@ -29,4 +29,16 @@
...
+ Initializes an MDCCurvedCornerTreatment instance with a given corner size.
+ */
+- (nonnull instancetype)initWithSize:(CGSize)size NS_DESIGNATED_INITIALIZER;
+
+/**
+ Initializes an MDCCurvedCornerTreatment instance with a corner size of zero.
+ */
+- (nonnull instancetype)init;
+
+- (nullable instancetype)initWithCoder:(nonnull NSCoder *)aDecoder NS_DESIGNATED_INITIALIZER;
```
Comment (jverkoey) [inline components/private/ShapeLibrary/src/MDCCurvedCornerTreatment.h]: Break this out to a separate PR if possible please.

File: components/private/Shapes/src/MDCShapedShadowLayer.h [hunk 3]
```diff
@@ -72,4 +72,12 @@
  */
 @property(nonatomic, strong, nonnull) CAShapeLayer *shapeLayer;
 
+/*
+ A sublayer of @c shapeLayer that is responsible for the background color of the shape layer.
+
+ The colorLayer imitates the path of shapeLayer and is added as a sublayer. It is updated when
+ shapedBackgroundColor is set on the layer.
+ */
+@property(nonatomic, strong, nonnull) CAShapeLayer *colorLayer;
```
Comment (jverkoey) [inline components/private/Shapes/src/MDCShapedShadowLayer.h]: Break this out to a separate PR if possible please.

File: components/Buttons/src/MDCButton.m (hunk 1 above)
Comment (yarneo) [reply components/Buttons/src/MDCButton.m]: I don't see why we would need to call super in this case, but please correct me if i'm unaware of something. What self.layer.shapedBackgroundColor does is call the sublayer's colorfill if there is a shape set, or otherwise call the self.layer.backgroundColor (which has the same effect as calling self.backgroundColor just without calling the setter itself of MDCButton). Because MDCButton subclasses UIButton and this call is called inside setBackgroundColor then i feel like the same flow remains.

File: components/private/ShapeLibrary/src/MDCCurvedCornerTreatment.h (hunk 2 above)
Comment (yarneo) [reply components/private/ShapeLibrary/src/MDCCurvedCornerTreatment.h]: https://github.com/material-components/material-components-ios/pull/3378

File: components/private/Shapes/src/MDCShapedShadowLayer.h (hunk 3 above)
Comment (yarneo) [reply components/private/Shapes/src/MDCShapedShadowLayer.h]: https://github.com/material-components/material-components-ios/pull/3379

**Comment by yarneo** [comment]
_Does "Stroked Triangular" intentionally not have ink?_

It did have ink only with very low opacity (look closely), i updated it to be the default ink again for more clarity.

File: components/Buttons/src/MDCButton.m (hunk 1 above)
Comment (jverkoey) [reply components/Buttons/src/MDCButton.m]: Ah got it! I can't tell from this PR whether the call to super is important, but if it looks like that behavior is expected here it might help to document that self's backgroundColor will eventually get set as expected.

Comment (yarneo) [reply components/Buttons/src/MDCButton.m]: added a comment to clarify behavior. yeah i thought the same, i had looked at the commit history though and it looks like super is called instead of self because the setBackgroundColor method was deprecated and therefore it would of been a no-op if we had called self.backgroundColor: https://github.com/material-components/material-components-ios/commit/c99b5d0014eee336949165b473b6202eb0e8eea6

File: components/Buttons/src/MDCButton.m [hunk 4]
```diff
@@ -923,6 +926,29 @@ - (void)updateTitleFont {
...
 }
 
+- (void)setShapeGenerator:(id<MDCShapeGenerating>)shapeGenerator {
+  if (shapeGenerator) {
+    self.layer.shadowPath = nil;
+  } else {
+    self.layer.shadowPath = [self boundingPath].CGPath;
+  }
+  self.layer.shapeGenerator = shapeGenerator;
+  self.layer.colorLayer.zPosition = -1;
```
Comment (jverkoey) [inline components/Buttons/src/MDCButton.m]: Please document why this is necessary.

---

Alternatively: can the color layer simply be pushed to the bottom of the layer hierarchy?

File: components/Buttons/src/MDCButton.h [hunk 5]
```diff
@@ -20,6 +20,8 @@
 #import "MaterialInk.h"
 #import "MaterialShadowElevations.h"
 
+@protocol MDCShapeGenerating;
```
Comment (jverkoey) [inline components/Buttons/src/MDCButton.h]: Import the shapes component that exposes this API instead please - we're still generally trying to avoid forward declarations of symbols not contained within a given component.

**Review by jverkoey — APPROVED** [review]
Looks good to me. Please confirm that you've tested other button demos in the catalog to verify that there was no regression in behavior or visual appearance.

File: components/Buttons/src/MDCButton.m (hunk 4 above)
Comment (yarneo) [reply components/Buttons/src/MDCButton.m]: Instead of setting the zPosition I had moved the colorLayer to be at its rightful place by removing and adding it below the inkView and the button's imageView. Also have written an explanation as a comment elaborating why

File: components/Buttons/src/MDCButton.h (hunk 5 above)
Comment (yarneo) [reply components/Buttons/src/MDCButton.h]: done

**Comment by yarneo** [comment]
Confirming that I have looked at other examples using buttons in the catalog and have not noticed any regressions or visual appearance changes to the best of my knowledge.
